docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+# Advanced Filtering for Triggers Proposal
+
+## Problem
+
+As an `Event Consumer` I want to filter events in triggers with a variety of
+strategies beyond exact match on predefined attributes. Filtering strategies
+that have been requested include:
+
+*   Filtering by source or type prefix or regular expression
+*   Filtering by a custom extension attribute
+*   Filtering by data fields
+
+The existing filter capability of exact match on source and type is simple to
+understand and use, but doesn't support these more advanced filtering scenarios.
+
+## Objective
+
+Design a flexible filtering mechanism that supports all of the advanced use
+cases identified and can also implement simpler cases like exact match.
+
+The filtering mechanism MUST be safe and efficient. It SHOULD be compatible with
+a future
+[upstream propagation solution](https://github.com/knative/eventing/issues/934)
+for delivery efficiency at scale.
+
+Filters should be easy for the user to specify and simple to interpret at a
+glance.
+
+## Requirements
+
+*   Support all of the advanced filtering use cases identified.
+*   Can implement simpler cases like exact match.
+*   Safe and secure when embedded in multi-tenant processes.
+*   Compatible with upstream filter propagation.
+*   Easily specified by the user.
+*   Simple for users to interpret.
```

Comment:
What about simple for users to create and interpret? Are there tools that will validate if the CEL expression is valid?

Reply:
CEL offers a type-checker which can ensure that all variables and functions available in the expression are declared and used with appropriate typing. When all the types are dynamic though (as is the case for JSON), static analysis is pretty limited.

Reply:
Good point @trshafer about validation feedback. I think we should add parsing to the webhook so that if the expression cannot be parsed or is invalid, it is rejected immediately. I'll add this to the proposal. It won't be perfect, as @TristonianJones points out, but it will catch some errors.